### ISSUE TITLE
Phaser assigns incorrect SNVs to genes

### ISSUE THREAD

**Issue by GrantHov** [secastel/phaser#32]
Dear Stephane,

I am experiencing some problems with Phaser results.
I need to obtain allelic counts for hybrid yeast, and I know the reference sequence of one of its parents.
My pipeline is following:
Generate bam files using STAR (sorted by coordinates) by mapping reads from hybrid to known reference parental.  Then generate vcf file on RNAseq data  using GATK pipeline (I skip base recalibration step), then I do filtering of the vcf file with parameters mentioned in GATK pipeline (http://gatkforums.broadinstitute.org/gatk/discussion/3891/calling-variants-in-rnaseq).

Then I bgzip and tabix fitered vcf file.

Here are some lines of final vcf file
NC_018292.1    1078    .    G    A    100.28    PASS    AC=2;AF=1.00;AN=2;DP=3;ExcessHet=3.0103;FS=0.000;MLEAC=2;MLEAF=1.00;MQ=60.00;QD=33.43;SOR=1.179    GT:AD:DP:GQ:PL    1/1:0,3:3:9:128,9,0
NC_018292.1    1130    .    T    C    100.28    PASS    AC=2;AF=1.00;AN=2;DP=3;ExcessHet=3.0103;FS=0.000;MLEAC=2;MLEAF=1.00;MQ=60.00;QD=33.43;SOR=1.179    GT:AD:DP:GQ:PL    1/1:0,3:3:9:128,9,0
NC_018292.1    1249    .    A    T    178.90    PASS    AC=2;AF=1.00;AN=2;DP=5;ExcessHet=3.0103;FS=0.000;MLEAC=2;MLEAF=1.00;MQ=60.00;QD=34.24;SOR=1.022    GT:AD:DP:GQ:PL    1/1:0,5:5:15:207,15,0
NC_018292.1    1369    .    T    C    107.28    SnpCluster    AC=2;AF=1.00;AN=2;DP=2;ExcessHet=3.0103;FS=0.000;MLEAC=2;MLEAF=1.00;MQ=60.00;QD=29.09;SOR=2.303    GT:AD:DP:GQ:PL    1/1:0,2:2:9:135,9,0

My command line for phaser is:
    python phaser/phaser/phaser.py --vcf filt_merged.vcf.gz --bam Aligned_sorted.bam \
--paired_end 1 --mapq 255 --baseq 10 --sample sample --id_separator "=" \
--threads 10 --o phaser_out

Then I do:
python phaser_gene_ae.py --haplotypic_counts phaser_out.haplotypic_counts.txt \
--features orth.bed --o phaser_gene_ae.txt --no_gw_phase 1

I generated .bed file form gff file just by selecting corresponding columns.

In gene output file I see strange behavour of phaser:
Many snps are sssigned to incorrect genes (you can see it by coordinates)
NC_018292.1     1048416 1052312 CORT_0A04800    3763    3200    6963    0.233811384293  61      NC_018292.1=948682=C=T,NC_018292.1=948822=T=C,NC_018292.1=948823=A=G,
NC_018292.1     1054556 1055605 CORT_0A04810    3763    3200    6963    0.233811384293  61      NC_018292.1=948682=C=T,NC_018292.1=948822=T=C,NC_018292.1=948823=A=G,
NC_018292.1     1056216 1056737 CORT_0A04820    3763    3200    6963    0.233811384293  61      NC_018292.1=948682=C=T,NC_018292.1=948822=T=C,NC_018292.1=948823=A=G,

I'd appreciate very much if you can give a hint on how to troubleshoot this issue.
Thank you,
Hrant

**Comment by secastel** [comment]
Hi there, indeed this is odd behavior. Would you mind uploading the "haplotypic_counts.txt" file along with the features .bed that you used so I can try to reproduce the issue?

**Comment by GrantHov** [comment]
Hi Stephane,
Thanks for quick reply.
Here you are. Had to change .bed to .txt to upload the bed file here.

[phaser_rep1.haplotypic_counts.txt](https://github.com/secastel/phaser/files/1116182/phaser_rep1.haplotypic_counts.txt)
[test_gene_counts.txt](https://github.com/secastel/phaser/files/1116184/test_gene_counts.txt)

[orth.bed.txt](https://github.com/secastel/phaser/files/1116186/orth.bed.txt)

**Comment by secastel** [comment]
Hi there, so I've spent some time looking into this and I have some insight.

The first issue is that in its current implementation phASER generates counts at the haplotype level. That is, it first generates haplotypes by connecting variants spanned by reads, and then it counts the number of reads that mapped to each of the two haplotypes. The way that phaser gene AE is currently implemented, is that it assigns haplotypic counts to each gene from any haplotypes that overlap it. Thus, a given gene can receive counts from a haplotype that spans the gene even if none of the individual variants found in that haplotype are found within the gene. This becomes a problem when for example, some transcripts are found in the introns of another transcript.

I've updated phaser gene AE so that now, for a gene to have a haplotype's reads assigned to it, at least one of the variants in that haplotype has to be physically found in the gene. In this case, you will still see variants reported in the gene ae output that are not themselves found in the gene, but they are part of a haplotype block constructed where one of the variants in it does overlap that gene. 

The best way to estimate haplotypic expression for individual genes would be to use only counts from variants that are directly found in the gene annotation. This change will be implemented in the next major release of phASER, which I am hoping to have out within the next month.

For applications in human RNA-seq data I don't expect making this change will affect a huge number of gene quantifications. I'm not exactly sure what you are looking for in your case, so it's hard for me to judge how much such a change would affect your results.

Sorry - I hope this isn't too confusing, and thanks for bringing this to my attention!

**Comment by GrantHov** [comment]
Hi Stephane,
Thanks for detailed explanation.
To be honest considering that for my data (yeasts) genes are quite linear (also there is almost no splicing and introns),  one would not expect to have many (even any) cases that you have mentioned, i.e. transcripts are found in introns of other transcripts. 
What I actually found is that apparently phaser sometimes wrongly defines the haplotype block. See and  example below.

    
    NC_018295.1	72477	72806	NC_018295.1=72477=C=T,NC_018295.1=72626=C=T,NC_018295.1=72650=T=C,NC_018295.1=72665=T=C,NC_018295.1=72806=A=T	5	T,T,C,C,T	C,C,T,T,A	41	35	76	0/1	0.5
    NC_018295.1	73291	73539	NC_018295.1=73291=T=C,NC_018295.1=73296=A=T,NC_018295.1=73400=G=A,NC_018295.1=73425=C=T,NC_018295.1=73458=A=G,NC_018295.1=73539=A=G	6	C,T,A,T,G,G	T,A,G,C,A,A	49	46	95	0/1	0.5
    NC_018295.1	76947	87384	NC_018295.1=76947=C=A,NC_018295.1=77092=G=A,NC_018295.1=77134=A=G,NC_018295.1=77155=G=A,NC_018295.1=77219=C=T,NC_018295.1=77499=G=A,NC_018295.1=77530=T=A,NC_018295.1=86852=C=A,NC_018295.1=86958=A=C,NC_018295.1=86975=T=C,NC_018295.1=86997=G=A,NC_018295.1=87018=A=G,NC_018295.1=87048=T=C,NC_018295.1=87057=G=A,NC_018295.1=87135=G=A,NC_018295.1=87285=C=T,NC_018295.1=87384=T=G	17	C,G,A,G,T,G,T,C,C,T,G,A,T,G,G,C,T	A,A,G,A,C,A,A,A,A,C,A,G,C,A,A,T,G	1647	789	2436	0/1	0.5
    NC_018295.1	77824	78182	NC_018295.1=77824=C=T,NC_018295.1=78182=A=T	2	T,T	C,A	5	31	36	0/1	0.5
    NC_018295.1	78583	78971	NC_018295.1=78583=C=A,NC_018295.1=78805=G=A,NC_018295.1=78814=C=T,NC_018295.1=78850=T=C,NC_018295.1=78895=A=G,NC_018295.1=78912=A=G,NC_018295.1=78971=C=T	7	C,G,C,T,A,A,C	A,A,T,C,G,G,T	569	149	718	0/1	0.5

I sorted haplotypes by coordinates, and I guess they have to be sequential (and in most of the cases they are), however you can see that there is one (third line) that jumps by 10kb, and that haplotype messes up hayplotypic counts for genes that are within that haplotype.
And you can observe the same behavioure for other regions as well.

**Comment by secastel** [comment]
I see where the confusion lies. In the example you have shown phASER hasn't actually done anything incorrectly. phASER assembles "haplotype blocks" not based on on contiguous coordinates, but using phase informative reads. Any two variants that are found on the same read will be assigned to the same haplotype block. In the case of RNA-seq reads, you can almost think of haplotype blocks as transcripts. Therefore, it is possible that some haplotype blocks are contained within others, as is the case with the example I gave of a transcript being within the intron of another gene. I've attached a diagram to help illustrate this.

This is the way phASER currently works. However, in the next update, which should be posted within the week, phASER will only use counts from variants that directly overlap the gene annotation, and not counts from any entire haplotype block that overlaps a gene annotation.

In short - if you wait about ~ 1 week, and try re-running the phASER on your data the results should be more straightforward for your use case.

![ink-image](https://user-images.githubusercontent.com/9827367/27883976-84063aaa-61a0-11e7-9e68-f1995562f46f.png)

**Comment by GrantHov** [comment]
Hi Stephane,

Thank you for your detailed clarifications.
Will wait for the new release then.

Kind regards,
Hrant

**Comment by GrantHov** [comment]
Hi Stephane,

Just a quick question. If one read contains several SNVs, then when counting allelic counts this read is counted once or it adds up for each SNV?

Thank you,
Hrant

**Comment by secastel** [comment]
If a read contains several SNVs it will only get counted once. This is to avoid double counting of reads, since they are not independent observations.

FYI finishing up final testing on the big update, expect to have it out by end of the week if all goes well.

**Comment by GrantHov** [comment]
Hi Stephane,

Thanks for quick reply!
Glad to hear that we will have the new release soon!

Cheers,
Hrant

**Comment by secastel** [comment]
New release is live. Make sure to update phaser.py and phaser_gene_ae.py, and give it a shot. Please let me know how it works for you, as there were a lot of changes, and I tried to do a lot of bug testing, but you never know!

**Comment by GrantHov** [comment]
Thanks a lot for you work! I will test it and get back to you as soon as possible!

Cheers,
Hrant
